File: 0725-split-linked-list-in-parts/0725-split-linked-list-in-parts.py
class Solution:
    def splitListToParts(self, head: Optional[ListNode], k: int) -> List[Optional[ListNode]]:
        result, size, temp = [None for i in range(k)], 0, head

        while temp:
            size += 1
            temp = temp.next

        chunk = size // k
        left = size % k

        for i in range(k):
            result[i] = head
            size_now = chunk + (1 if left > 0 else 0)
            left -= 1

            for j in range(size_now - 1):
                if head:
                    head = head.next

            if head:
                temp = head.next
                head.next = None
                head = temp

        return result

File: 0725-split-linked-list-in-parts/0725-split-linked-list-in-parts.cpp
class Solution {
public:
    vector<ListNode*> splitListToParts(ListNode* head, int k) {
        vector<ListNode*> result(k);
        ListNode* temp = head;
        int size = 0;

        while (temp != NULL) {
            ++size;
            temp = temp->next;
        }

        int chunk = size / k, left = size % k;

        for (int i = 0; i < k; i++) {
            result[i] = head;
            int size_now = chunk + (left > 0 ? 1 : 0);
            --left;

            for (int j = 0; j < size_now - 1; j++) {
                if (head != NULL) {
                    head = head->next;
                }
            }

            if (head != NULL) {
                temp = head->next;
                head->next = NULL;
                head = temp;
            }
        }

        return result;
    }
};